src/celeritas/global/CoreParams.cc

```diff
@@ -87,23 +87,29 @@ CoreParams::CoreParams(Input input) : input_(std::move(input))
     CP_VALIDATE_INPUT(rng);
     CP_VALIDATE_INPUT(init);
     CP_VALIDATE_INPUT(action_reg);
+    CP_VALIDATE_INPUT(max_streams);
 #undef CP_VALIDATE_INPUT
 
     CELER_EXPECT(input_);
 
+    CoreScalars scalars;
```

Comment:
This could be a potential usecase of IIFE:
```cpp
    auto const scalars = [this]() -> CoreScalars {
        // Construct geometry action
        const auto boundary_action{input_.action_reg->next_id()};
        input_.action_reg->insert(
            std::make_shared<celeritas::generated::BoundaryAction>(
                boundary_action, "geo-boundary", "Boundary crossing"));

        // Construct implicit limit for propagator pausing midstep
        const auto propagation_limit_action{input_.action_reg->next_id()};
        input_.action_reg->insert(std::make_shared<ImplicitGeometryAction>(
            propagation_limit_action,
            "geo-propagation-limit",
            "Propagation substep/range limit"));

        // Save maximum number of streams
        const auto max_streams{input_.max_streams};

        return {boundary_action, propagation_limit_action, max_streams};
    }();
```

Reply:
I think the use of initializers here reduces readability and increases the chance of an error: if someone switches the order of the boundary/propagation in the struct this will cause a nasty subtle bug. Also, one of the main reasons I like IIFE is to limit the scope of temporaries that are just used to calculate another parameter; and the original code doesn't have any temporaries (it saves directly to the scalars).

Reply:
That's fair, while we don't have temporaries here I guess other advantages are that you can declare scalars const and I would say it also clearly outlines the code initializing scalars. This could be a more readable, less error-prone middle-ground:

```cpp
    auto const scalars = [this]() {
        CoreScalars scalars;
        // Construct geometry action
        scalars.boundary_action = input_.action_reg->next_id();
        input_.action_reg->insert(
            std::make_shared<celeritas::generated::BoundaryAction>(
                scalars.boundary_action, "geo-boundary", "Boundary crossing"));

        // Construct implicit limit for propagator pausing midstep
        scalars.propagation_limit_action = input_.action_reg->next_id();
        input_.action_reg->insert(std::make_shared<ImplicitGeometryAction>(
            scalars.propagation_limit_action,
            "geo-propagation-limit",
            "Propagation substep/range limit"));

        // Construct action for killed looping tracks
        scalars.abandon_looping_action = input_.action_reg->next_id();
        input_.action_reg->insert(
            std::make_shared<ImplicitSimAction>(scalars.abandon_looping_action,
                                                "abandon-looping",
                                                "Abandoned looping track"));

        // Save maximum number of streams
        scalars.max_streams = input_.max_streams;
        return scalars;
    }();
```